src/filters/BallFilter.cpp

```diff
@@ -0,0 +1,139 @@
+        CameraFilter(detectTime, cameraID),
+        lastPredictTime{detectTime} {
+    KalmanInit(detectionBall);
+}
+void BallFilter::KalmanInit(const proto::SSL_DetectionBall &detectionBall) {
+    // SSL units are in mm, we do everything in SI units.
+    double x = mmToM(detectionBall.x());//m
+    double y = mmToM(detectionBall.y());//m
+    Kalman::Vector startState = {
+            x, y, 0, 0
+    };
+
+    Kalman::Matrix startCov;
+    startCov.eye();
+    //initial noise estimates
+    const double startPosNoise = 0.05;
+    startCov.at(0, 0) = startPosNoise;//m noise in x
+    startCov.at(1, 1) = startPosNoise;//m noise in y
+
+    kalman = std::make_unique<Kalman>(startState, startCov);
+
+    kalman->H.eye();     // Our observations are simply what we see.
+
+
+}
+void BallFilter::applyObservation(const proto::SSL_DetectionBall &detectionBall, int cameraID) {
+    Kalman::VectorO observation = {mmToM(detectionBall.x()), mmToM(detectionBall.y())};
+    //TODO: do things with the other ball fields (pixel pos, area)
+    kalman->z = observation;
+
+    //Observations which are not from the main camera are added but are seen as much more noisy
+    const double posVar = 0.02; //variance TODO: tune these 2
+    const double posVarOtherCamera = 0.05;
+    kalman->R.zeros();
+    if (cameraID == mainCamera) {
+        kalman->R.at(0, 0) = posVar;
+        kalman->R.at(1, 1) = posVar;
+    } else {
+        kalman->R.at(0, 0) = posVarOtherCamera;
+        kalman->R.at(1, 1) = posVarOtherCamera;
+    }
+    kalman->update();
+}
+proto::WorldBall BallFilter::asWorldBall() const {
+    proto::WorldBall msg;
+    const Kalman::Vector &state = kalman->state();
+    msg.mutable_pos()->set_x(state[0]);
+    msg.mutable_pos()->set_y(state[1]);
+    msg.mutable_vel()->set_x(state[2]);
+    msg.mutable_vel()->set_y(state[3]);
+    msg.set_visible(ballIsVisible());
+    //TODO: add height filter here, and actually set the z, z_vel, area fields
+    return msg;
+}
+double BallFilter::distanceTo(double x, double y) const {
+    const Kalman::Vector &state = kalman->state();
+    double dx = state[0] - mmToM(x);
+    double dy = state[1] - mmToM(y);
+    return sqrt(dx * dx + dy * dy);
+}
+void BallFilter::predict(double time, bool permanentUpdate, bool cameraSwitched) {
+    double dt = time - lastUpdateTime;
+    // forward model:
+    kalman->F.eye();
+    kalman->F.at(0, 2) = dt;
+    kalman->F.at(1, 3) = dt;
+
+    //TODO: add 2 stage forward model?
+    //Set B
+    kalman->B = kalman->F;
+    //Set u (we have no control input at the moment)
+    kalman->u.zeros();
+
+    //Set Q matrix
+    const float posNoise = 0.1;//TODO: tune
+    Kalman::MatrixO G;
+    G.zeros();
+    G.at(0, 0) = dt * posNoise;
+    G.at(0, 2) = 1 * posNoise;
+    G.at(1, 1) = dt * posNoise;
+    G.at(1, 3) = 1 * posNoise;
+    if (cameraSwitched){
+        G.at(0,0)+=0.05;
+        G.at(1,1)+=0.05;
+    }
+    kalman->Q = G.t() * G ;
```

Comment:
I am not sure what this computation does, it looks kind of strange. Why is Q made by G^T * G?

Reply:
The answer to this and your above question is that this is a model. You essentially model Q=G^t G, which is equivalent to saying that you model the process noise as normally  distributed with mean 0 and average. The current model here is quite bad but is what we used before.  See https://en.wikipedia.org/wiki/Kalman_filter for another application if you're curious. Koen knows more about this than I do probably.

Reply:
Yeah the G^T G makes sense now, I am used to it being given directly :smile: However from what you told me I was thinking there would be some adjustment made to the R matrix if the camera that the observation comes from is not preferred